File: app/src/main/java/com/codingblocks/restapiretrofitjson/adapters/PhotosAdapter.java
package com.codingblocks.restapiretrofitjson.adapters;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.codingblocks.restapiretrofitjson.R;
import com.codingblocks.restapiretrofitjson.interfaces.OnItemClickListener;
import com.codingblocks.restapiretrofitjson.models.Photo;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;

/**
 * Created by GhanshamBansal on 06/07/17.
 */

public class PhotosAdapter extends RecyclerView.Adapter<PhotosAdapter.PhotosViewHolder> {

    private Context context;
    private ArrayList<Photo> photos;
    private OnItemClickListener onItemClickListener;

    public PhotosAdapter(Context context, ArrayList<Photo> photos) {
        this.context = context;
        this.photos = photos;
    }

    public void setOnItemClickListener(OnItemClickListener onItemClickListener)
    {
        this.onItemClickListener = onItemClickListener;
    }

    public void updatePhotos(ArrayList<Photo> photos){
        this.photos = photos;
        notifyDataSetChanged();
    }

    @Override
    public PhotosViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater li = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        View itemView = li.inflate(R.layout.list_item_photos,parent,false);
        return new PhotosViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(PhotosViewHolder holder, int position) {
        final Photo thisPhoto = photos.get(position);
        holder.tvPhoto.setText(thisPhoto.getTitle());
        Picasso.with(context).load(thisPhoto.getThumbnailUrl()).into(holder.image);
        holder.thisView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onItemClickListener.onItemClick(thisPhoto.getId(),v);
            }
        });


    }

    @Override
    public int getItemCount() {
        return photos.size();
    }

    class PhotosViewHolder extends RecyclerView.ViewHolder{
        TextView tvPhoto;
        ImageView image;
        View thisView;
        public PhotosViewHolder(View itemView) {
            super(itemView);
            thisView = itemView;
            image = (ImageView) itemView.findViewById(R.id.imageView);
            tvPhoto = (TextView) itemView.findViewById(R.id.tv_Photo);

        }
    }
}

File: app/src/main/java/com/codingblocks/restapiretrofitjson/adapters/PhotosAdapter.kt
package com.codingblocks.restapiretrofitjson.adapters


import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import com.codingblocks.restapiretrofitjson.R
import com.codingblocks.restapiretrofitjson.interfaces.OnPhotoClickListener
import com.codingblocks.restapiretrofitjson.models.Photos
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.list_item_photos.view.*
import java.util.*

/**
 * Created by pervy_sage on 30/6/17.
 */

class PhotosAdapter(internal var photos: ArrayList<Photos>, internal var mContext: Context) : RecyclerView.Adapter<PhotosAdapter.PhotosViewHolder>() {

    internal var onPhotoClickListener: OnPhotoClickListener? = null
    fun setOnPhotoClickListener(onPhotoClickListener: OnPhotoClickListener) {
        this.onPhotoClickListener = onPhotoClickListener
    }

    fun updateAlbums(photos: ArrayList<Photos>, progressBar: ProgressBar) {
        this.photos = photos
        progressBar.visibility = View.INVISIBLE
        progressBar.visibility = View.GONE
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PhotosViewHolder {
        val li = mContext.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        val itemView = li.inflate(R.layout.list_item_photos, parent, false)
        return PhotosViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: PhotosViewHolder, position: Int) {
        holder.bindPhoto(photos[position])
    }

    override fun getItemCount(): Int {
        return photos.size
    }

    inner class PhotosViewHolder(internal var rootView: View)
        : RecyclerView.ViewHolder(rootView) {
        fun bindPhoto(thisPhoto : Photos){
            rootView.tvTitle.text=thisPhoto.title
            Picasso.with(mContext)
                    .load(thisPhoto.thumbnailUrl)
                    .error(R.drawable.loading)
                    .placeholder(R.drawable.loading)
                    .into(rootView.ivThumbnail)
            rootView.setOnClickListener {
                onPhotoClickListener!!.onClick(thisPhoto.thumbnailUrl,thisPhoto.title)
            }
        }
    }
}
